doc/reset_factory/ResetFactoryHLD.md

```diff
@@ -0,0 +1,236 @@
+
+4. It should restore "/etc/sonic" directory to factory folder (Clear the directory in overlayfs upperdir).
+except SONiC immutable variables file "sonic-environment"
+
+5. It should reboot the system to load and apply the new configuration.
+
+6. It should stop sonic.target to ensure that files are not deleted while other daemons are utilizing them.
+
+7. it should allow users/application to create their own factory default configurations or
+extend the reset factory functionality.
+
+
+####  1.5.2. <a name='ConfigurationManagementRequirements'></a>Configuration and Management Requirements
+We will create a command line tool /usr/bin/reset-factory which should provide following functionality:
+
+1. default: reset configurations to factory default, clear all system logs and files.
+
+2. keep-basic: preserves basic configurations only after boot. Logs and files will be deleted except users history files and home directories.
+
+3. keep-all-config: preserves all configurations after boot. Logs and files will be deleted.
+
+4. only-config: reset configurations to factory default. Logs and files will be preserved.
+
+    
+##  2 <a name='Design'></a>Design
+	
+###  2.1 <a name='High-LevelDesign'></a>High-Level Design
+![diagram](diagram.png)
+
+(flow description in the chapter below)
+
+"reset-factory" script will create new config_db.json if needed using "config-setup factory" command.<br/>
+In addition, it will clear logs, tech-support and reboot-cause files, users history files and home directories.<br/>
+
+Steps:
+
+1. Stops sonic.target: 
+* monit unmonitor container_checker
+* systemctl stop sonic.target
+
+2. If not "keep-all-config":
+
+config-setup factory <None/keep-basic/only-config>
+
+3. If not ("only-config" or “keep-basic”):
+
+* Delete all non-default users and restore default passwords of default users.
+* Delete bash, vim and python history files under "/home".
+* Delete any non-dotfiles in the home directories.
+
+4. If not "only-config":
+* Remove all docker containers
+* Restore "/etc/sonic" directory to factory folder (Clear the directory in overlayfs upperdir)
+* Delete all files under "/host/warmboot"
+* Delete tech-support files under "/var/dump/"
+* Delete all files under "/var/log"
+* Delete all files and symlinks under "/host/reboot-cause/"
+
+5. Print to /var/log/systemlog:
+New log file that won't be cleared in reset-factory.
+It will allow user to track operations such as "reset-factory".
+
+6. Reboot
+
+###  2.2 <a name='UsersDefaults'></a>Users Defaults
+
+During build, we will create a new file "/etc/sonic/default_users.json" that will store
+default user accounts names and default hashed passwords.
+The file will have same permissions as /etc/shadow "-rw-------" (600).
+```
+{
+    "admin": {
+        "password": "$y$j9T$qGjjyhXSHtuHwowV3DE44.$wrDK6/vecdYIE8ujXyVPkGHvlDHP.ySCsBTLvOdg.31"
+    }
+}
+```
+We will remove all non-default user accounts (e.g. 1000 < UID <= 6000),
```

Comment:
You intention is to restore the /etc/shadow to the original one at build time. I propose to mount/host/image-CURRENT/fs.squashfs, and copy the original file to the current file system.

Reply:
Thanks for the suggestion
I think it might be better to do the following ( similar to /etc/sonic clear step):
1. rm -f /host/image-CURRENT/rw/etc/shadow
2. mount -o remount /

I will test it and update the code accordingly

Reply:
After investigating it more deeply, I found out that modifying the shadow file directly isn't recommended at all
The recommended way to update these file is through the official API commands ( useradd, usermod...)
In addition, it is not enough to restore shadow file only, we also should restore /etc/passwd, /etc/group
(probably some other files also)
I believe the current approach is much safer.